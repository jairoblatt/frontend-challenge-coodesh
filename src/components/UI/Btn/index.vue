<template>
  <button v-on="$listeners" :class="classes">
    <spinner v-show="loading" size="21" />
    <slot v-show="!loading" />
  </button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";

export default Vue.extend({
  name: "Btn",

  props: {
    primary: Boolean as PropType<boolean>,
    secondary: Boolean as PropType<boolean>,
    block: Boolean as PropType<boolean>,
    loading: Boolean as PropType<boolean>,
    small: Boolean as PropType<boolean>,
    outlined: Boolean as PropType<boolean>,
  },

  computed: {
    classes() {
      return {
        btn: true,
        "btn--primary": this.primary,
        "btn--secondary": this.secondary,
        "btn--block": this.block,
        "btn--small": this.small,
        "btn--outlined": this.outlined,
      };
    },
  },
});
</script>

<style lang="postcss" scoped>
.btn {
  @apply rounded-md px-5 py-3 text-sm outline-none font-semibold text-white transition-colors;
}

.btn--primary {
  @apply bg-primary-base hover:bg-primary-light;
}

.btn--secondary {
  @apply bg-secondary-gray;
}

.btn--block {
  @apply w-full;
}

.btn:disabled {
  @apply bg-gray-300 text-gray-900 cursor-auto;
}

.btn--small {
  @apply px-7 py-2;
}

.btn--outlined {
  @apply border border-gray-100 text-light-onSurfacePrimary hover:border-black hover:text-black;
}
</style>
