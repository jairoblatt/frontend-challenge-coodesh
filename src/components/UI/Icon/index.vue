<template>
  <i v-if="iconSanitize" :class="`mdi ${iconSanitize}`" :style="styles"></i>
</template>

<script lang="ts">
import Vue, { PropOptions, PropType } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Icon",

  props: {
    size: {
      type: [String, Number],
      default: 14,
    } as PropOptions<string | number>,

    color: String as PropType<string>,

    icon: {
      type: String,
      default: "",
    } as PropOptions<string>,
  },

  computed: {
    iconSanitize(): string {
      return this.$slots.default?.[0]?.text || this.icon;
    },

    styles(): CSS.Properties {
      return {
        fontSize: this.size + "px",
        ...(this.color && { color: this.color }),
      };
    },
  },
});
</script>
