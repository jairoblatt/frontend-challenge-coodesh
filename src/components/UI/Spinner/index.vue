<template>
  <div class="spinner-container">
    <span class="spinner" :style="styles"></span>
  </div>
</template>
<script lang="ts">
import Vue, { PropOptions } from "vue";
import CSS from "csstype";

export default Vue.extend({
  name: "Spinner",

  props: {
    size: {
      type: [String, Number],
      default: "inherit",
    } as PropOptions<string | number>,
  },

  computed: {
    styles(): CSS.Properties {
      return {
        minHeight: `${this.size}px`,
        minWidth: `${this.size}px`,
      };
    },
  },
});
</script>
<style lang="postcss" scoped>
.spinner-container {
  @apply relative flex justify-center items-center min-h-full min-w-full;
}

.spinner {
  @apply pointer-events-none border-solid rounded-full  border-primary-base  p-1;

  border-width: 0.3rem;
  border-top-color: white;
  animation: spinner 0.5s linear infinite;
}

@keyframes spinner {
  100% {
    transform: rotate(360deg);
  }
}
</style>
