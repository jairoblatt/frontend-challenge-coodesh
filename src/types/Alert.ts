export interface Alert {
  id?: string | number;
  timeOut?: number | string | boolean;
  title: string;
  description?: string;
  theme: "success" | "warn" | "fail" | "default";
}
