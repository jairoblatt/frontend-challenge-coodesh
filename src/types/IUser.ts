export type Nullable<T> = T | null;

export interface UserApiInfo {
  page: number;
  results: string;
  seed: string;
  version: string | number;
  gender?: string;
  nat?: string;
}

export interface Location {
  city: string;
  coordinates: {
    latitude: string;
    longitude: string;
  };
  country: string;
  postcode: number;
  state: string;
  street: {
    number: number;
    name: string;
  };
  timezone: {
    offset: string;
    description: string;
  };
}

export interface UserName {
  title: string;
  first: string;
  last: string;
}

export interface Login {
  md5: string;
  password: string;
  salt: string;
  sha1: string;
  sha256: string;
  username: string;
  uuid: string;
}
export interface UserData {
  cell: string;
  email: string;
  gender: string;
  id: { name: string; value: string };
  location: Location;
  name: UserName;
  login: Login;
  nat: string;
  phone: string;
  picture: {
    large: string;
    medium: string;
    thumbnail: string;
  };
  dob: {
    date: string;
    age: number;
  };
}

export interface Users {
  results: Nullable<UserData[]>;
  info: Nullable<UserApiInfo>;
}

export interface Filters {
  name: any;
  gender: string;
  nat: string;
}
