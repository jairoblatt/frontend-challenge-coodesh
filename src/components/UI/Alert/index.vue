<template>
  <div :class="['alert', `alert--${content.theme}`]">
    <Btn class="alert__head" @click="removeAlert">
      <Icon size="20" icon="mdi-close" />
    </Btn>
    <div class="alert__content">
      <span>{{ content.title }}</span>
      <p v-if="content.description">{{ content.description }}</p>
    </div>
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from "vue";
import { IAlert } from "@/types/IAlert";
import { sleep } from "@/utils";

export default Vue.extend({
  name: "Alert",

  props: {
    content: {
      type: Object,
      required: true,
      default: () => null,
    } as PropOptions<IAlert>,
  },

  async mounted() {
    if (this.content.timeOut) {
      await sleep(Number(this.content.timeOut));
      this.removeAlert();
    }
  },

  methods: {
    removeAlert() {
      this.$store.dispatch("alert/remove", this.content.id);
    },
  },
});
</script>
<style lang="postcss" scoped>
.alert {
  @apply flex flex-col items-center shadow-md right-2 md:right-11 md:translate-x-px top-8 fixed bg-white px-2 md:px-5 rounded-sm max-w-sm md:max-w-md;

  border-left: 7px solid;
  z-index: 999999;
}

.alert--success {
  @apply border-primary-base;
}

.alert--wran {
  border-color: #fb8c00;
}

.alert--fail {
  border-color: #ff1744;
}

.alert--default {
  border-color: transparent;
}

.alert__head {
  @apply self-end cursor-pointer absolute top-0 right-3 p-2 text-light-onSurfacePrimary;
}

.alert__content {
  @apply pt-1 pb-4;
}

.alert__content span {
  @apply text-light-onSurfacePrimary font-bold text-lg pb-0;
}

.alert__content p {
  @apply text-light-onSurfacePrimary text-base;
}
</style>
