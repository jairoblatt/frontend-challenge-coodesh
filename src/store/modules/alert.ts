import { GetterTree, MutationTree, ActionTree } from "vuex";
import { IAlert } from "@/types/IAlert";

export const state = () => ({
  alerts: [] as IAlert[],
});

export type AlertState = ReturnType<typeof state>;

export const getters: GetterTree<AlertState, AlertState> = {
  get: (state) => state.alerts[0] || null,
};

export const mutations: MutationTree<AlertState> = {
  SET_ALERT(state, payload) {
    state.alerts.push(payload);
  },

  UPDATE_ALERT(state, payload) {
    state.alerts = payload;
  },
};

export const actions: ActionTree<AlertState, AlertState> = {
  set({ commit }, payload: IAlert) {
    const { timeOut } = payload;
    commit("SET_ALERT", {
      ...payload,
      timeOut: timeOut === false ? timeOut : Number(timeOut) > 1 ? timeOut : 10000,
      id: Math.random()
        .toString(36)
        .substr(2),
    });
  },

  remove({ commit, state }, alertId) {
    commit(
      "UPDATE_ALERT",
      state.alerts.filter(({ id }) => id !== alertId)
    );
  },
};
