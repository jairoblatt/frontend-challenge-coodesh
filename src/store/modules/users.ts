import { GetterTree, MutationTree, ActionTree } from "vuex";
import HTTPClient from "@/plugins/axios";
import { Nullable, Users, UserApiInfo } from "@/types/IUser";
import { sleep } from "@/utils";

export const state = () => ({
  users: null as Nullable<Users[]>,
  fetchInfo: null as Nullable<UserApiInfo>,
  fetchPending: false,
});

export type UserState = ReturnType<typeof state>;

export const getters: GetterTree<UserState, UserState> = {
  get: (state) => state.users,
  fetchInfo: (state) => state.fetchInfo,
  fetchPending: (state) => state.fetchPending,
};

export const mutations: MutationTree<UserState> = {
  SET_USERS(state, payload) {
    state.users = payload;
  },

  FETCH_PENDING(state, status) {
    state.fetchPending = status;
  },

  FETCH_INFO(state, info) {
    state.fetchInfo = info;
  },
};

export const actions: ActionTree<UserState, UserState> = {
  async fetch({ commit, state }, context) {
    commit("FETCH_PENDING", true);
    try {
      const gender = context?.gender;
      const nat = context?.nat;

      const { data } = await HTTPClient.get<Users>("/", {
        params: {
          results: 50,
          seed: context?.seed,
          page: context?.page || 1,
          nat,
          gender,
        },
      });

      if (data?.results) {
        commit("SET_USERS", [...(state.users || []), ...data.results]);
      }

      if (data?.info) {
        commit("FETCH_INFO", {
          ...data.info,
          ...(gender && { gender }),
          ...(nat && { nat }),
        });
      }
    } catch {
      commit("FETCH_INFO", null);
    } finally {
      await sleep(400);
      commit("FETCH_PENDING", false);
    }
  },
};
