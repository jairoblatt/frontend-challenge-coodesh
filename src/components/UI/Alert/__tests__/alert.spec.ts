import { mount, Wrapper, MountOptions, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import * as AlertModule from "@/store/modules/alert";
import AlertComponent from "../index.vue";
import Icon from "@/components/UI/Icon/index.vue";
import Btn from "@/components/UI/Btn/index.vue";
import Spinner from "@/components/UI/Spinner/index.vue";
import { Alert } from "@/types/Alert";

describe("Alert.vue ", () => {
  const localVue = createLocalVue();
  localVue.component("Icon", Icon);
  localVue.component("Btn", Btn);
  localVue.component("Spinner", Spinner);
  localVue.use(Vuex);

  const removeAlertSpy = jest.spyOn((AlertComponent as any).options.methods, "removeAlert");

  type Instance = InstanceType<typeof AlertComponent>;
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>;

  const alertContent: Alert = {
    id: "123",
    timeOut: 1000,
    title: "My title",
    description: "My description",
    theme: "fail",
  };

  beforeEach(() => {
    mountFunction = (options = {}) => {
      const store = new Vuex.Store({
        modules: {
          alert: {
            namespaced: true,
            ...AlertModule,
          },
        },
      });

      return mount(AlertComponent, {
        propsData: {
          content: alertContent,
        },
        ...options,
        store,
        localVue,
      });
    };
  });

  it("Should mount the component", () => {
    const wrapper = mountFunction();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should apply the current alert theme", () => {
    const wrapper = mountFunction();
    expect(wrapper.classes("alert--" + alertContent.theme)).toBe(true);
  });

  it('Should call the dispatch "alert/remove" when Btn is clicked', () => {
    const wrapper = mountFunction();

    wrapper.findComponent(Btn).trigger("click");

    expect(removeAlertSpy).toHaveBeenCalled();
  });

  it('Should render the "title" and "description" when content prop is passed', () => {
    const wrapper = mountFunction();

    const spanElement = wrapper.findAll("span").at(1);
    const pElement = wrapper.find("p");

    expect(spanElement.text()).toBe(alertContent.title);
    expect(pElement.text()).toBe(alertContent.description);
  });

  it('Should have "mdi-close" into btn', () => {
    const wrapper = mountFunction();

    const iconComponent = wrapper.findComponent(Icon);

    expect(iconComponent.classes()).toEqual(["mdi", "mdi-close"]);
  });
});
