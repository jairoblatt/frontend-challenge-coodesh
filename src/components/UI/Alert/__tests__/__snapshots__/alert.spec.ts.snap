// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert.vue  Should mount the component 1`] = `
"<div class=\\"alert alert--fail\\"><button class=\\"alert__head btn\\">
    <div class=\\"spinner-container\\" style=\\"display: none;\\"><span class=\\"spinner\\" style=\\"min-height: 21px; min-width: 21px;\\"></span></div> <i class=\\"mdi mdi-close\\" style=\\"font-size: 20px;\\"></i>
  </button>
  <div class=\\"alert__content\\"><span>My title</span>
    <p>My description</p>
  </div>
</div>"
`;
